### DISCUSSION TITLE
Question about parsing terminals in a text string

### DISCUSSION THREAD

**Discussion by mnaderhirn** [eclipse-langium/langium#1573] [Q&A] [answered]
Good day.

I have this grammar:
```
grammar TitleGrammar

import 'common.tokens';

entry Title:
    'TITLE:' title=Text;

grammar CommonTokensGrammar

import 'release.tokens'

entry CommonTokens:
    token = INDENT | DEDENT | NAME | Text;


//hidden terminal NEWLINE: 'this_string_does_not_matter_newline#$%^&*((';

terminal NEWLINE: /[\r\n]+/;
terminal DEDENT: 'this_string_does_not_matter_dedent#$%^&*((';
terminal INDENT: 'this_string_does_not_matter_indent#$%^&*((';
//hidden terminal SPACES: ' ';
hidden terminal SPACES:/\s+/;

THE returns string: 'THE'|'The'|'the';
SHALL returns string: 'SHALL'|'Shall'|'shall';
terminal COMMA: /,/;
terminal DOT: /\./;

terminal TRIPLE_QUOTE: /"""/;
terminal COLON: /:/;
terminal QUOTE: /'/;


hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;

terminal NAME: /[_a-zA-Z][\w_]*/;
Keyword returns string: THE | SHALL;
Text returns string: (Keyword | NAME)+;
```
I have implemented a Lexer to parse Indent and Dedent, which works just fine. Now i want to recognize sentences like
The bird is here. For this I have defined 
- NAME as terminal to parse general words
- terminals like THE or SHALL should be included 
- Text is the string i want to return
if i start it in VS Code i get for this error from the language server:

Request initialize failed with message: Errors detected in definition of Lexer:
Token Type: ->NAME<- static 'GROUP' can only be Lexer.SKIPPED/Lexer.NA/A String

I really dont know what the issue is here. What be great if some could give me some hints 

Thank you,
Michael

**Comment by msujew** [comment]
Hey @mnaderhirn,

> I have implemented a Lexer to parse Indent and Dedent, which works just fine. Now i want to recognize sentences like

So you're saying you created a specialized `TokenBuilder` service? Can you share that? I have reason to believe a minor implementation detail might lead to the error.

**Reply by mnaderhirn** [reply]
```ts
import _ from 'lodash';
import { DefaultTokenBuilder, LexerResult, TokenBuilderOptions, DefaultLexer, GrammarUtils } from 'langium';
import { LangiumCoreServices } from 'langium';
import { Grammar, GrammarAST } from 'langium';
import { createToken, createTokenInstance, IToken, TokenType, Lexer } from 'chevrotain';
import { } from './generated/ast.js';

let indentStack = [0];

function matchIndentBase(text: string, offset: number, matchedTokens: IToken[], groups: any, type: string): RegExpExecArray | null {
    const isStartOfLine = offset === 0 || text[offset - 1] === '\n' || text[offset - 1] === '\r';

    if (isStartOfLine) {
        let match: RegExpExecArray | null;
        let currIndentLevel: number | undefined = undefined;
        const wsRegExp = /[ \t]+/y;
        wsRegExp.lastIndex = offset;
        match = wsRegExp.exec(text);

        if (match !== null) {
            currIndentLevel = match[0].length;
        } else {
            currIndentLevel = 0;
        }

        const prevIndentLevel = _.last(indentStack);
        if (prevIndentLevel === undefined) {
            throw Error();
        }

        if (currIndentLevel > prevIndentLevel && type === 'indent') {
            indentStack.push(currIndentLevel);
            //console.log('Indent found! Indent level: ', indentStack, 'match: ', match?.[0]);
            matchedTokens.push(
                createTokenInstance(Indent, "INDENT", NaN, NaN, NaN, NaN, NaN, NaN)
            );
            return null; //since match contains \t it is better to use createTakenInstance
        } else if (currIndentLevel < prevIndentLevel && type === 'dedent') {
            const matchIndentIndex = _.findLastIndex(
                indentStack,
                (stackIndentDepth) => stackIndentDepth === currIndentLevel
            );

            if (matchIndentIndex === -1) {
                throw Error(`invalid outdent at offset: ${offset}`);
            }

            const numberOfDedents = indentStack.length - matchIndentIndex - 1;

            let iStart = match !== null ? 1 : 0;
            for (let i = iStart; i < numberOfDedents; i++) {
                indentStack.pop();
                matchedTokens.push(
                    createTokenInstance(Dedent, "DEDENT", NaN, NaN, NaN, NaN, NaN, NaN)
                );
            }

            if (iStart === 1) {
                indentStack.pop();
            }
            return match;
        } else {
            return null;
        }
    } else {
        return null;
    } 
}

const matchIndent = _.partialRight(matchIndentBase, "indent");
const matchOutdent = _.partialRight(matchIndentBase, "dedent");

const Indent = createToken({ name: "INDENT", pattern: matchIndent, line_breaks: false });
const Dedent = createToken({ name: "DEDENT", pattern: matchOutdent, line_breaks: false });
//const Name = createToken({ name: "NAME", pattern: /[a-zA-Z_][\w]*/, group: 'Names'});
//const Newline = createToken({ name: "NEWLINE", pattern: /\r?\n/, group: Lexer.SKIPPED });

export class EasysafeLexer extends DefaultLexer {
    constructor(services: LangiumCoreServices) {
        super(services);
    }

    override tokenize(text: string): LexerResult {
        console.log("Start Tokenize of text: ", text);
        let chevrotainResult = super.tokenize(text);
        console.log("Tokens:", chevrotainResult.tokens.map(t => t.image).join(", "));
        let newlines = text.split(/\r\n|\r|\n/).length;
        //console.log('newlines: ', newlines);
        let match = text.match(/.*$/)?.length;
        match = match ? match : 0;

        while (indentStack.length > 1) {
            //console.log('Adding remaining DEDENT token');
            chevrotainResult.tokens.push(
                createTokenInstance(Dedent, "DEDENT", text.length, text.length, newlines, newlines, match, match)
            );
            indentStack.pop();
        }

        indentStack = [0];

        chevrotainResult.tokens = chevrotainResult.tokens.filter(token => token.tokenType.name !== 'SPACES');

        console.log("Finished Tokenize!");
        return chevrotainResult;
    }
}

export class EasysafeTokenBuilder extends DefaultTokenBuilder {
    override buildTokens(grammar: Grammar, options?: TokenBuilderOptions): TokenType[] {
        console.log('buildTokens!!!!');
        let tokens: TokenType[] = super.buildTokens(grammar, options) as TokenType[];

        let updatedTokens: TokenType[] = [];

        for (const token of tokens) {
            if (token.name !== "SPACES") {
                updatedTokens.push(token);
            }
        }

        console.log('Updated Tokens:', updatedTokens.map(t => t.name));
        return updatedTokens;
    }

    protected override buildTerminalToken(terminal: GrammarAST.TerminalRule): TokenType {
        let group: string | undefined;
        const regex = GrammarUtils.terminalRegex(terminal);
        console.log('buildTerminalToken for:', terminal.name);

        if (terminal.hidden) {
            group = new RegExp(regex).test(' ') ? Lexer.SKIPPED : 'hidden';
        }
        

        let token;
        if (terminal.name === "NEWLINE") {
            token = createToken({
                name: "NEWLINE",
                pattern: /\r?\n/,
                group: Lexer.SKIPPED
            });
        } else if (terminal.name === "INDENT") {
            //console.log('Indent token creation');
            token = createToken({
                name: "INDENT",
                pattern: matchIndent,
                line_breaks: false
            });
        } else if (terminal.name === "DEDENT") {
            token = createToken({
                name: "DEDENT",
                pattern: matchOutdent,
                line_breaks: false
            });
        } else if (terminal.name === "SPACES") {
            token = createToken({
                name: "SPACES",
                pattern: / +/,
                group: Lexer.SKIPPED
            });
        }
        else {
            console.log('no name in terminal.name: ', terminal.name, ' hahahah: ', group);
            token = createToken({
                name: terminal.name,
                pattern: new RegExp(regex),
                group: group
            });
        }

        return token;
    }
}
```

**Reply by msujew** [reply]
Alright, thanks. In here:

```ts
token = createToken({
  name: terminal.name,
  pattern: new RegExp(regex),
  group: group
});
```

You possibly set `group` to undefined. That's actually not allowed. Chevrotain will complain (and you've seen that first-hand). See our workaround for this:

https://github.com/eclipse-langium/langium/blob/721b882c8f2e9eecfcc0ba445876ef1c4dd1ed21/packages/langium/src/parser/token-builder.ts#L58-L61

**Reply by mnaderhirn** [reply]
} else if (terminal.name === "NAME") {
            token = createToken({
                name: "NAME",
                pattern: /[_a-zA-Z][\w_]*/,
                group: 'Text'
            });
 edit this line, but now a line like, compiler error disappears:
Now if i right if type in VS Code this line:
The bird is here.
I get this errors:
[{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "Expecting: one of these possible Token sequences:\n  1. [NAME]\n  2. [TITLE:]\n  3. [[RELEASE_DOCUMENT]]\nbut found: 'The'",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 4
},{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "Expecting end of file but found `The`.",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 4
},{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "unexpected character: -> <- at offset: 3, skipped 1 characters.",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 4,
	"endLineNumber": 1,
	"endColumn": 5
},{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "unexpected character: -> <- at offset: 8, skipped 1 characters.",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 9,
	"endLineNumber": 1,
	"endColumn": 10
},{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "unexpected character: -> <- at offset: 11, skipped 1 characters.",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 12,
	"endLineNumber": 1,
	"endColumn": 13
},{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "unexpected character: ->.<- at offset: 16, skipped 1 characters.",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 17,
	"endLineNumber": 1,
	"endColumn": 18
}]
the Name Token is not recognized, the The token which i dont understand

Oh of course to you the indent and dedent if it is of any use!

**Reply by msujew** [reply]
Giving a token a `group` removes it from the token stream. This is only useful for tokens that you do not want to consume in your parser (like comments or whitespace). See also [this doc](https://chevrotain.io/docs/features/token_grouping.html#token-grouping).

**Reply by mnaderhirn** [reply]
Thank you for your help.
terminal NAME: /[_a-zA-Z][\w_]*/;  this is the definition of NAME. Since it is not hidden wont jump into 
if (terminal.hidden) {
            group = new RegExp(regex).test(' ') ? Lexer.SKIPPED : 'hidden';
        }
What should i do instead?
Lexer.NA?

**Reply by msujew** [reply]
Just don't use a group at all. It's enough to define it as:

```ts
token = createToken({
  name: 'NAME',
  pattern: /some-regex/
});
```

**Reply by mnaderhirn** [reply]
Looks better now, thank you so far:
Now typing: The bird is here in VS Code results to 
```json
[{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "Expecting: one of these possible Token sequences:\n  1. [NAME]\n  2. [TITLE:]\n  3. [[RELEASE_DOCUMENT]]\nbut found: 'The'",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 4
},{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "Expecting end of file but found `The`.",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 4
},{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "unexpected character: -> <- at offset: 3, skipped 1 characters.",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 4,
	"endLineNumber": 1,
	"endColumn": 5
},{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "unexpected character: -> <- at offset: 8, skipped 1 characters.",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 9,
	"endLineNumber": 1,
	"endColumn": 10
},{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "unexpected character: -> <- at offset: 11, skipped 1 characters.",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 12,
	"endLineNumber": 1,
	"endColumn": 13
}]
```

**Reply by msujew** [reply]
1. Why are you removing `SPACES` from the token list in your token builder? It should still be lexed, but hidden. Your code simply removes the token type, which prevents the lexer from reading whitespaces.
2. How does your grammar look like? Do you allow parsing of `The` as the first token?

**Reply by mnaderhirn** [reply]
@msujew -- Thank you for being so helpful
made the spaces not hidden as you can see in my grammar and that resolved the space erros, but now i am getting this errors for The
The bird is here --> see errors below the grammar. The is a keyword and should be tokenized since i use it for a formal language
```
grammar CommonTokensGrammar

import 'release.tokens'

entry CommonTokens:
    token = INDENT | DEDENT | NAME | Text;


//hidden terminal NEWLINE: 'this_string_does_not_matter_newline#$%^&*((';

terminal NEWLINE: /[\r\n]+/;
terminal DEDENT: 'this_string_does_not_matter_dedent#$%^&*((';
terminal INDENT: 'this_string_does_not_matter_indent#$%^&*((';
//hidden terminal SPACES: ' ';
terminal SPACES:/\s+/;

THE returns string: 'THE'|'The'|'the';
SHALL returns string: 'SHALL'|'Shall'|'shall';
terminal COMMA: /,/;
terminal DOT: /\./;

terminal TRIPLE_QUOTE: /"""/;
terminal COLON: /:/;
terminal QUOTE: /'/;


hidden terminal ML_COMMENT: /\/\*[\s\S]*?\*\//;
hidden terminal SL_COMMENT: /\/\/[^\n\r]*/;

terminal NAME: /[_a-zA-Z][\w_]*/;
Keyword returns string: THE | SHALL;
Textpart returns string: (Keyword | NAME);
Text: text+=Textpart SPACES* (text+=Textpart)*;
```

```json
[{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "Expecting: one of these possible Token sequences:\n  1. [NAME]\n  2. [TITLE:]\n  3. [[RELEASE_DOCUMENT]]\nbut found: 'The'",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 4
},{
	"resource": "/home/michael/projects/test_project/test.easy",
	"owner": "_generated_diagnostic_collection_name_#0",
	"severity": 8,
	"message": "Expecting end of file but found `The`.",
	"source": "easysafe",
	"startLineNumber": 1,
	"startColumn": 1,
	"endLineNumber": 1,
	"endColumn": 4
}]
```

**Reply by msujew** [reply, marked as answer]
I've adjusted your grammar a bit - the `Text` rule didn't allow any spaces after the second word. See [this playground grammar](https://langium.org/playground?grammar=OYJwhgthYgBAwgeyogdgFUQawKaoM4Di4UMAUGXgC4gCeCaV1AXLAG5gA2ArjgLzocADyoBuCmQD0kgBYBLACYK8sJiAhzUXWADkAogHUAMgEl9rAORV5%2BAPr4am4LYWIcd1Iiq3oVNbdQcAHdOTRwAYgASAFIAPQAyACoACmSLcTI1DS1OXUNTc1hJAG0AHRBS1ABdAGpJcSzNbQARPVaddEtrOTsHECcXNw8vHzA-HBAXHGVUKii4pNT0zIns7TN2ztgrG3tHVGdXdwCR339NGbmYhJS08Wl5JRVGnNgAZQAFAEF4PTfLbYNVZNXKfH5-ZiSUr4OoZdAACT0sBAOCo3BABFgfSclgRegsAB8LOgZDhCTsyeI3vCvkYjMjUejMdiDpZqbSjOS3jIuJxyfgeZw%2BUD1CCEAB5ACykq%2BrEkABp6itRa9muKtlCAHRK5VrXLoABKJg%2BRj0tgAigBVdV6OUAIgdSpe2ng4qM4p0cshIr1sCtNrlFh1ZEeM1UwNekqMtld0r0HTlpShiTK%2BFKbyqiQA-KVEkmlaHnhHtG9o7HJfGNUmk8VYpVypmdc7cjovhW5cVbGAALQALy%2B3YAWlUykFbI3xABpHC0IKIEAKBlojH4LH7YCsPGwAnvGl08SCEQABxgVCXTNXLI3sGS09n88XO9bFYAlAfhFRWEwRDUBB%2BTyAZ7JGCvxvIk4Y-n%2Bx6ni%2BiTiEAA&content=CoCwpgBARglgTgEwjAzhccxA).

**Reply by mnaderhirn** [reply]
Yep now it is working. Thank you!
Cool project!

**Reply by msujew** [reply]
Thanks, you're welcome :)

**Reply by mnaderhirn** [reply]
Request initialize failed with message: Errors detected in definition of Lexer:
Token Type: ->NAME<- static 'GROUP' can only be Lexer.SKIPPED/Lexer.NA/A String

What excatly is causing this errors?

**Reply by msujew** [reply]
@mnaderhirn See [this code in Chevrotain](https://github.com/Chevrotain/chevrotain/blob/4272462f2c88846357dc95dfdd5684d87c153c4f/packages/chevrotain/src/scan/lexer.ts#L690-L714). The `GROUP` field of a token type cannot be set to `undefined`.

**Comment by mnaderhirn** [comment]
Thank you:

grammar DateGrammar

entry Datum:
  'Date:' year=YEAR ('-'|'.') month=MONTH ('-'|'.') day=DAY;

terminal YEAR: /20[0-9]{2}/;
terminal MONTH: /1[0-2]|[1-9]/;
terminal DAY: /3[01]|[12][0-9]|[1-9]/;

see [playground](https://langium.org/playground?grammar=OYJwhgthYgBAImALgUwOLijAUNlA7JEATwWQFcIAubWWAckVSvtmJRgF4BNAUQEEASrAAU9ALT0APvQB09AJSwIAe0IALTgFkA8gDkAKgAlRE6XMWwAJmGKd4-bgG5cqEBACW%2BMABtYfISpYAHoAJgAGAG1w8QBOAF0Ab1CAX2CXN09vP11DIyDggEZo8VD4qUjCuPj07EyvXwRHAoBmaMLyyrKShIqqhNrsIA&content=CIQwLgpgXATADAgtARhooA)

Maybe a very primitive question: 
How do i have to organize this, so that i can differentiate betweent Month and Day?

**Reply by msujew** [reply]
Hey @mnaderhirn,

Short answer: You can't, since the tokenization phase cannot take context into account.

Long answer: I would recommend using just a single `INTEGER` terminal and then using a validation to ensure that the values of `year`, `month` and `day` fall within the expected ranges. Not only does this solve the parsing issue, but also improve the comprehension of the error messages; While trying to solve this using tokenization leads to some pretty obscure and hard to read error messages in the editor, using validations allows to build specific error messages for your use case. The general design philosophy for building DSLs is usually: Be flexible in what you accept and create specific and useful error/warning messages for invalid inputs instead of rejecting them via the parser.

**Reply by mnaderhirn** [reply]
Understood, Thank you

**Comment by mnaderhirn** [comment]
[Example](https://langium.org/playground?grammar=OYJwhgthYgBAagUxAZwJYHsB2AoPisAXEATwWXWwC4dY7YByeAUQCUBlASQHkA5KhrHYAFAIIBhZuwD8sAG7SAvAzkNZ0AFYYQizrwAqjAHSCIaLNt0HYACgYnYABzCEAxgAsr%2BgJTSA3Hj0sITIZlhgADawvMwA6gAyesxUsAD0ANoAOiDSmVgAugDUqQEhIGGRsAAizDUGKQyE7mgoAPooxObArQAmGIhtFoSt0IRlvYg9BIQAxAAkAKQAegBkAFQ2dqWh5pV6dfoNTS3tnVjdfQOtQyMu4%2BZTRPPL65sMAampzT2PwTvhUREEikDUY23Ku0BYkk7CoqUyKGKARwZQqUT0hhAiEIAFcQFgULAsDiIAAjZApDIABgAtABOIolPCoyHRUQAWWSaXSrTANIAXqIaQAtfJZADurXyayZQA&content=GoUQSgygkg8gcgLgAQGYB0BWTQ)

This grammar seems to work perfectly in the playground, but when i run a test case
it('should parse the version correctly', async () => {
        const input = 'VERSION: v12.3.6';
        const result = await parse(input);

        expect(result.parseResult).toBeDefined();
        const titleNode = result.parseResult.value;
        expect(titleNode).toBeDefined();
        expect(titleNode.major).toEqual(12);
        expect(titleNode.minor).toEqual(3);
        expect(titleNode.patch).toEqual(6);
    });

, I ll get:
v12 to be of token type v12 which makes sense. how should i adopt the grammar in order to ensure it is working right?
E. g. allowing 'v' to be a NAME token and then writing a validation function which returns an error?

**Reply by msujew** [reply]
I can't really follow. What is the error? If you want to use `v` as a name, you can follow [this guide](https://langium.org/docs/recipes/keywords-as-identifiers/).

Also, please consider opening separate discussions for separate questions. Makes my job a bit easier and allows for better indexing by search engines.
